### PR TITLE
Make CWallet::ListCoins atomic

### PR DESCRIPTION
**Pull request by promag**
Fix a potencial race in `CWallet::ListCoins`.

Replaces `cs_main` and `cs_wallet` locks by assertions in `CWallet::AvailableCoins`.

### CODE REVIEW

**Comment by MarcoFalke** [comment]
Just noting that you can fixup right now. It is possible to ignore-all-space in git diff

**Comment by promag** [comment]
I know but IIRC it's not possible to add comments with that option, so I've decided to split for now.

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -2198,111 +2198,109 @@ CAmount CWallet::GetAvailableBalance(const CCoinControl* coinControl) const
 
 void CWallet::AvailableCoins(std::vector<COutput> &vCoins, bool fOnlySafe, const CCoinControl *coinControl, const CAmount &nMinimumAmount, const CAmount &nMaximumAmount, const CAmount &nMinimumSumAmount, const uint64_t nMaximumCount, const int nMinDepth, const int nMaxDepth) const
 {
+    AssertLockHeld(cs_main);
+    AssertLockHeld(cs_wallet);
```
Comment (jamesob) [inline src/wallet/wallet.cpp]: Both are recursive locks, so any reason we can't just use `LOCK2` here?

Comment (promag) [reply src/wallet/wallet.cpp]: Long term plan is to remove recursive locks. Correct me if I'm wrong @TheBlueMatt.

Comment (ryanofsky) [reply src/wallet/wallet.cpp]: I think in general we want to move away from recursive locking: https://github.com/bitcoin/bitcoin/blob/85123be78df5a4e9447e70f9ee727416d45843d7/src/sync.h#L92

http://www.zaval.org/resources/library/butenhof1.html has some interesting rationale and history

Comment (promag) [reply src/wallet/wallet.cpp]: Thanks @ryanofsky for the reference.

Comment (jamesob) [reply src/wallet/wallet.cpp]: Ah, thanks for the clarification.

**Comment by devrandom** [comment]
ACK https://github.com/bitcoin/bitcoin/pull/12333/commits/834529f3c097386b82feb01bd2cc3e0f90418813

**Review by ryanofsky — COMMENTED** [review]
utACK 834529f3c097386b82feb01bd2cc3e0f90418813. To make the locking change easier to evaluate, I might give the second commit (834529f3c097386b82feb01bd2cc3e0f90418813) a description indicating it is an indentation-only change, instead of squashing it.

**Comment by sipa** [comment]
utACK 834529f3c097386b82feb01bd2cc3e0f90418813 (very easy to review with `-w` / `?w=1`).

**Comment by promag** [comment]
Followed @ryanofsky suggestion, [1st commit](https://github.com/bitcoin/bitcoin/pull/12333/commits/dc0f32252334b7c305e33c1e1151b0f445ce0ce3) and [2nd commit](https://github.com/bitcoin/bitcoin/pull/12333/commits/4848f093c4755794ca57c6514cb1e3c484cf3867?w=1).

**Comment by ryanofsky** [comment]
utACK 4848f093c4755794ca57c6514cb1e3c484cf3867, only change since last review is commit description

**Comment by promag** [comment]
Conflicts with #10605.

File: src/wallet/test/wallet_tests.cpp [hunk 2]
```diff
@@ -677,11 +677,11 @@ BOOST_FIXTURE_TEST_CASE(ListCoins, ListCoinsTestingSetup)
 
     // Lock both coins. Confirm number of available coins drops to 0.
     std::vector<COutput> available;
+    LOCK2(cs_main, wallet->cs_wallet);
```
Comment (laanwj) [inline src/wallet/test/wallet_tests.cpp]: This locks the cs_main and cs_wallet for the entire rest of this test, which is somewhat awkward scoping-wise, is that intentional?

Comment (promag) [reply src/wallet/test/wallet_tests.cpp]: Nothing else happens and below the lock is again required so I've decided for the minimal diff. I can lock only until `ListCoins` below if you prefer.

Comment (laanwj) [reply src/wallet/test/wallet_tests.cpp]: Yes please, For readability I prefer locks to be taken at the beginning of a block.

Comment (promag) [reply src/wallet/test/wallet_tests.cpp]: Done, 2 blocks instead of one.

**Pull request merged**